### PR TITLE
Support for TCP upgrade

### LINKED ISSUE #1123: The api client should send connection upgrade headers
To hint proxies about connection hijacking, docker clients should send connection upgrade headers like the docker cli does.

From https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/4-2-hijacking:

> In this version of the API, /attach, uses hijacking to transport stdin, stdout, and stderr on the same socket.
>
> To hint potential proxies about connection hijacking, Docker client sends connection upgrade headers similarly to websocket.

```
Upgrade: tcp
Connection: Upgrade
```

On Docker for Desktop, the proxy that sits between Docker Compose and the daemon will not be aware that the connection will be hijacked. This can lead to an issue where the proxy will install a CloseNotifier and just after that will hijack the connection, which is know to be incompatible. See https://github.com/docker/compose/issues/3685

See also https://github.com/docker/compose/issues/3700

### PR DESCRIPTION
**Pull request by aanand**
- Carried @dgageot's changes from #1124 
- Carried @shin-'s changes from #1126 
- Refactored common code between `docker/client.py` and `tests/helpers.py` into `docker/utils/socket.py`
- Cleaned up and documented socket methods

Fixes #1123 

ping @dnephin

### CODE REVIEW

File: docker/api/exec_api.py [hunk 1]
```diff
@@ -66,10 +64,18 @@ def exec_start(self, exec_id, detach=False, tty=False, stream=False,
...
+            'Connection': 'Upgrade',
+            'Upgrade': 'tcp'
+        }
+
         res = self._post_json(
-            self._url('/exec/{0}/start', exec_id), data=data, stream=stream
+            self._url('/exec/{0}/start', exec_id),
+            headers=headers,
+            data=data,
+            stream=True
```
Comment (dnephin) [inline docker/api/exec_api.py]: `stream` changed here to always `True`. Is that correct?

File: docker/utils/socket.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    return data
+
+
+def next_frame_size(socket):
+    """
+    Returns the size of the next frame of data waiting to be read from socket,
+    according to the protocol defined here:
+
+    https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/attach-to-a-container
+    """
```
Comment (dnephin) [inline docker/utils/socket.py]: Well that's fun

**Comment by dnephin** [comment]
code LGTM, waiting on janky

File: docker/api/exec_api.py (hunk 1 above)
Comment (shin-) [inline docker/api/exec_api.py]: Late answer, but yes, that's intentional. Without it, `_read_from_socket` cannot work properly. As far as the user is concerned, the output will still be presented according to their choice.

**Pull request merged**